### PR TITLE
tests: twister: using statemachine for status

### PR DESCRIPTION
**Pull request by hakehuang**
using statuemachine in status change.
in stastus.py we can generated the statuemachine
by get_graph

The purposes of this PR:

1. use statuemachine transition to verify the status change in twister runtime. if there is a violate, `TransitionNotAllowed` exception will be triggered, and call context will be print.
2. the transition is defined in the statues.py with class `TwisterStatusMachine_inner.cycles`, you can generate the status.png by run `python3 ./statues.py`
3. with this, we can see and control the status flow to avoid mis-use.
4. enable call stack print if a exception is triggered, in error.py
5. fix some status error found by this.

### CODE REVIEW

**Comment by hakehuang** [comment]
state
with this, we can observe the support test state transition as above

**Comment by golowanow** [comment]
>  with this, we can observe the support test state transition as above

is there already differentiation on status transitions for Harness, TestCase, TestSuite, and TestInstance ?

**Comment by hakehuang** [comment]
> is there already differentiation on status transitions for Harness, TestCase, TestSuite, and TestInstance ?

No, but I need clean the ci failures

**Comment by nashif** [comment]
> > is there already differentiation on status transitions for Harness, TestCase, TestSuite, and TestInstance ?
> 
> No, but I need clean the ci failures. BTW do you happen to know why this change does not get execute? [zephyrproject-rtos/zephyr/actions/runs/11608284181/workflow?pr=80623#L103](https://github.com/zephyrproject-rtos/zephyr/actions/runs/11608284181/workflow?pr=80623#L103)

because we are running in pull_request_target mode.

**Comment by hakehuang** [comment]
> because we are running in pull_request_target mode.

then how can I change this? @nashif

File: scripts/pylib/twister/twisterlib/statuses.py [hunk 1]
```diff
@@ -51,3 +66,215 @@ def get_color(status: TwisterStatus) -> str:
...
+        | block_statue.to.itself(internal=True,  cond="is_blocked")
+        | error_statue.to.itself(internal=True,  cond="is_error")
+        | notrun_statue.to.itself(internal=True, cond="is_force")
+        | filter_statue.to.itself(internal=True, cond="is_filtered")
+    )
+
+    def is_start(self, cond: TwisterStatus):
+        if cond == TwisterStatus.STARTED:
+            return True
+        return False
```
Comment (pdgendt) [inline scripts/pylib/twister/twisterlib/statuses.py:122]: Why use the `if` statement?
```suggestion
    def is_start(self, cond: TwisterStatus):
        return cond == TwisterStatus.STARTED
```

**Comment by PerMac** [comment]
@hakehuang to elaborate: pull_request_target means that the source is taken from the target pranch, not the PR. So your change has to be already in the tree to take place, which afaik doesn't allow to test the change itself within the PR. But you can split the workflow change from the rest and marge it to your github fork and then see if it works. Providing a link to a PR on a fork with the workflow change included  and working could help with the review. 
I think we also have repo made for testing https://github.com/zephyrproject-rtos/zephyr-testing but I never used it. We can discuss how to proceed such changes during testing wg.

**Comment by PerMac** [comment]
To reviewers: this PR requires manual testing before being approved

**Comment by PerMac** [comment]
@hakehuang Please provide more details.

**Comment by hakehuang** [comment]
> is there already differentiation on status transitions for Harness, TestCase, TestSuite, and TestInstance ?

@golowanow , from my observation, the Harness, TestCase, TestSuite and TestInstance transtions can follow one rule set. do you see any exceptions?

**Comment by hakehuang** [comment]
> Please provide more details.

The purposes of this PR:

1. use statuemachine transition to verify the status change in twister runtime. if there is a violate, `TransitionNotAllowed` exception will be triggered, and call context will be print.
2. the transition is defined in the statues.py with class `TwisterStatusMachine_inner.cycles`, you can generate the status.png by run `python3 ./statues.py`
3. with this, we can see and control the status flow to avoid mis-use.
4. enable call stack print if a exception is triggered, in error.py
5. fix some status error found by this.

**Comment by hakehuang** [comment]
> only some states/transitions valid for Harness, blocked is TestCase only, etc.

I see, let me add harness transition

File: scripts/pylib/twister/twisterlib/error.py
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/error.py:1]: ok, let me separate them

File: scripts/pylib/twister/twisterlib/statuses.py [hunk 2]
```diff
@@ -51,3 +66,185 @@ def get_color(status: TwisterStatus) -> str:
...
+
+    def trigger(self, event : TwisterStatus) -> None:
+        tsm = TwisterStatusMachine_inner()
+        tsm.force_state(self.current_state)
+        try:
+            tsm.trigger(event)
+        except exceptions.TransitionNotAllowed:
+            # here we violate the pre-defined transition rules
+            self.force_state(event)
+            g()
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/statuses.py:402]: yes, this is what we want to track, see below dump
```
WARNING - ======call stack dump start==========
INFO    - File "/usr/lib/python3.10/threading.py", line 973, in _bootstrap
    self._bootstrap_inner()
INFO    - File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
INFO    - File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/handlers.py", line 225, in _output_handler
    harness.handle(stripped_line)
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/harness.py", line 339, in handle
    tc.status = TwisterStatus.PASS
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/testsuite.py", line 381, in status
    self.case_status_machine.trigger(value)
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/statuses.py", line 200, in trigger
    g()
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/statuses.py", line 25, in g
    for line in traceback.format_stack():
WARNING - ======call stack dump end============
```
logger.exception can not dump the call stack, but g() does.

as you can see above,

```
INFO    - File "/home/shared/disk/zephyr_project/zephyr_test/zephyr/scripts/pylib/twister/twisterlib/harness.py", line 339, in handle
    tc.status = TwisterStatus.PASS
```
I need enable self assignment in harness

[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/statuses.py:402]: I find the issues is that harness.py line 341, when the testcase is load, it was set to Fail first.
```
tc.status = TwisterStatus.FAIL
```
actually it should be set as harness status
```
tc.status = self.status
```
@gchwier  now it works for me with your test command line

**Comment by hakehuang** [comment]
now this PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/81703

**Comment by hakehuang** [comment]
to pass CI, we need run
```
pip3 install python-statemachine
```

**Comment by hakehuang** [comment]
@nashif , any suggestion on how to integrate the python-statemachine in CI?

**Comment by PerMac** [comment]
Can you show how a status can:
can a transition from: none -> started -> filtered really happen? Sounds impossible. If it is actually the case, this is a point for the proper state transition evaluation and mapping map

**Comment by golowanow** [comment]
> with this, we can observe the support test state transition as above

would be nice to draw the state transition diagram vertically for better readability of its branches

**Comment by golowanow** [comment]
@hakehuang as an idea: to make the state machine a pluggable module (harness) to have it active only for Twister black box testing ?

File: scripts/pylib/twister/twisterlib/harness.py [hunk 3]
```diff
@@ -344,7 +344,7 @@ def handle(self, line):
         if self.status == TwisterStatus.PASS:
             tc.status = TwisterStatus.PASS
         else:
-            tc.status = TwisterStatus.FAIL
+            tc.status = self.status
```
[the remark this answers is not in the document]
Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/harness.py:347]: @golowanow, the case can be skipped, not FAIL.

[the remark this answers is not in the document]
Comment (golowanow) [reply scripts/pylib/twister/twisterlib/harness.py:347]: the reason to set to FAIL is described here:
https://github.com/zephyrproject-rtos/zephyr/blob/67a433fa048705444a46e1d1871dca6c89e41f75/scripts/pylib/twister/twisterlib/harness.py#L319-L325

**Comment by hakehuang** [comment]
> @hakehuang as an idea: to make the state machine a pluggable module (harness) to have it active only for Twister black box testing ?

let me change to use a parameter to enable it.

**Comment by hakehuang** [comment]
> > with this, we can observe the support test state transition as above
> 
> would be nice to draw the state transition diagram vertically for better readability of its branches

yes, Let me simplify the diagram

File: scripts/pylib/twister/twisterlib/statuses.py [hunk 4]
```diff
@@ -52,3 +68,384 @@ def get_color(status: TwisterStatus) -> str:
     FAIL = 'failed'
     PASS = 'passed'
     SKIP = 'skipped'
+
+
+class TwisterStatusMachineInner(StateMachine):
+    '''
+        twister status statue machine
+    '''
+    none_statue = State(str(TwisterStatus.NONE), initial=True)
```
Comment (nashif) [inline scripts/pylib/twister/twisterlib/statuses.py:77]: statue? Shouldn't this be 'status'?

Comment (hakehuang) [reply scripts/pylib/twister/twisterlib/statuses.py:77]: should be state, this is the concept of pystatemachine, let me update this.

File: scripts/pylib/twister/twisterlib/harness.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (golowanow) [reply scripts/pylib/twister/twisterlib/harness.py:347]: I think it is correct to set TC status to FAIL here either as a default status for any console line received while there is no clear PASS yet, or translating any harness non-Pass to TC failure. Assigning from harness.status might not harm as well (needs testing), but it is mostly just to keep it with same None here.

> Transition from passed to passed is not allowed ```, it is because the FAIL status is wrongly set by default.

IMO in practice it is ok to allow both PASS->PASS, or set FAIL by default.

**Comment by hakehuang** [comment]
> Can you show how a status can: can a transition from: none -> started -> filtered really happen? Sounds impossible. If it is actually the case, this is a point for the proper state transition evaluation and mapping map. Is the current map 1:1 based on the code or assumptions of how it works?

none-> started -> filtered is not happen by now, but it is possible. as for now, we marked all tests as SKIPED, which means not supported, but this could possible changed to FILTERED in command line. for me, 
SKIP means do not support
FILTERRED means we skip this in purpose.

the current map is implement is open to discussion, as you can see from the map, this mapping pass the unit test and blackbox testing. and it is a combination on code and some assumptions.

**Comment by PerMac** [comment]
> none-> started -> filtered is not happen by now, but it is possible. as for now, we marked all tests as SKIPED, which means not supported, but this could possible changed to FILTERED in command line. for me,

IMO the picture has to correspond to what we have effectively right now in the software. First, you propose the addition of the state machine corresponding to the current state. When this idea is accepted we can evaluate transitions and propose changes to them. If not, I am worried this PR will be an endless discussion of "do this proposed transition make sense or not" instead of focusing if the tool itself should be used.

**Comment by hakehuang** [comment]
> IMO the picture has to correspond to what we have effectively right now in the software. First, you propose the addition of the state machine corresponding to the current state. When this idea is accepted we can evaluate transitions and propose changes to them. If not, I am worried this PR will be an endless discussion of "do this proposed transition make sense or not" instead of focusing if the tool itself should be used.

@PerMac  thanks for the comments, I remove the not existing transitions. and update the diagram as well.
